other/perun-propagate/etc/perun_propagate

```diff
@@ -39,41 +52,116 @@ check() {
 }
 
 perun_propagate_start() {
+	HOST_NOT_FOUND="doesn't have assigned any facility";
 	CC_FILE=`mktemp`
 	trap "unlink ${CC_FILE}" ERR
 	export KRB5CCNAME="FILE:${CC_FILE}"
 
+	#choose the type of kerberos and prepare Principals (all unique hosts)
 	set -e
 	if ktutil --version </dev/null 2>&1 | grep -qi heimdal; then
 		# Heimdal
-		PRINCIPAL=`ktutil list 2>/dev/null | awk '($3~"^host/") && ($3!~"\\\.ipv6\\\.") { print $3 }' | head -1`
-		kinit --no-afslog --keytab="${KEYTAB}" "${PRINCIPAL}"
+		PRINCIPALS=`ktutil list 2>/dev/null | awk '($3~"^host/") && ($3!~"\\\.ipv6\\\.") { print $3 }' 2>/dev/null | sort | uniq`
+		KERBEROS_TYPE="heimdal"
 	else
 		# MIT Kerberos
-		PRINCIPAL=`echo -e "read_kt ${KEYTAB}\nlist" | ktutil list 2>/dev/null | awk '($3~"^host/") && ($3!~"\\\.ipv6\\\.") { print $3 }' | head -1`
-		kinit -k -t "${KEYTAB}" "${PRINCIPAL}"
+		PRINCIPALS=`echo -e "read_kt ${KEYTAB}\nlist" | ktutil list 2>/dev/null | awk '($3~"^host/") && ($3!~"\\\.ipv6\\\.") { print $3 }' 2>/dev/null | sort | uniq `
+		KERBEROS_TYPE="mit"
 	fi
 
+	#For every unique Principal do kinit and start perun-propagate
 	set +e
-	if which timeout stdbuf &>/dev/null; then
-		OUT=`timeout ${TIMEOUT} stdbuf -oL -eL remctl "${PERUN_SERVER}" perun propagate 2>&1`
-		EXIT_CODE=$?
-	else
-		OUT=`remctl "${PERUN_SERVER}" perun propagate 2>&1`
-		EXIT_CODE=$?
+	ITERATOR=0;
+	while read -r LINE; do
+		if [ "${KERBEROS_TYPE}" = "heimdal" ]; then
+			kinit -k -t "${KEYTAB}" "${LINE}"
+		else	
+			kinit --no-afslog --keytab="${KEYTAB}" "${LINE}"
+		fi
+		TMP_ERROR=`mktemp perun-propagate.XXXXXXXXX`
+		remctl "${PERUN_SERVER}" perun propagate 2>&1 | tee ${TMP_ERROR} >> ${LOG} &
+		PID=$!
+		PID_ARRAY[$ITERATOR]=$PID
+		TMP_ERROR_ARRAY[$ITERATOR]=$TMP_ERROR
+		PRINCIPAL_ARRAY[$ITERATOR]=$LINE
+	done <<< "$PRINCIPALS"
+
+	#Start time of propagation in seconds
+	START_TIME=`date +%s`
+	#Prepare firt run of while with same START_TIME
+	NOW_TIME=$START_TIME
+	#Count of all processes in PIDS_ARRAY to know how much of them still running
+	PROCESSES_COUNT=${#PID_ARRAY[@]}
+	#Waiting for all running processes but not more than TIMEOUT in seconds
+	while [ `expr $NOW_TIME - $START_TIME` -lt "$TIMEOUT" ]; do
+		DONE_PROCESSES=0
+		for i in ${!PID_ARRAY[*]}; do
+			if is_process_running ${PID_ARRAY[$i]} ; then
+				break
```

Comment:
Consider adding short sleep before this "break". This script don't have to consume 100% processor time.